src/docker-extract.c

```diff
@@ -0,0 +1,342 @@
+#define _GNU_SOURCE
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <errno.h>
+#include <string.h>
+#include <fcntl.h>
+#include <sys/file.h>
+#include <sys/stat.h>
+#include <archive.h>
+#include <archive_entry.h>
+
+#include "config.h"
+#include "util/file.h"
+#include "util/message.h"
+#include "util/registry.h"
+#include "util/util.h"
+
+/* apply_opaque
+ *  Given opq_marker as a path to a whiteout opaque marker
+ *    e.g. usr/share/doc/test/.wh..wh..opq
+ *  Make the directory containing the make opaque for this layer by removing it
+ *  if it exists under rootfs_dir
+ */
+int apply_opaque(const char *opq_marker, char *rootfs_dir) {
+    int retval = 0;
+    char *token, *opq_dir, *opq_dir_rootfs;
+    size_t buff_len;
+
+    token = strrchr(opq_marker, '/');
+
+    if (token == NULL) {
+        singularity_message(ERROR, "Error getting dirname for opaque marker\n");
+        ABORT(255);
+    }
+
```

Comment:
To simplify and because that code doesn't run multi-threaded, I would define a static buffer for string manipulation and I would make sure snprintf() is able to NULL terminate the string. Here's what you can do instead : 

int apply_opaque(const char *opq_marker, char *rootfs_dir) {                                                                 
    int retval;                                                                                                              
    char *token;                                                                                                             
    static char buf[MAXPATHLEN];                                                                                             
                                                                                                                             
    token = strrchr(opq_marker, '/');                                                                                        
    if (token == NULL) {                                                                                                     
        singularity_message(ERROR, "Error getting dirname for opaque marker\n");                                             
        ABORT(255);                                                                                                          
    }                                                                                                                        
                                                                                                                             
    retval = snprintf(buf, sizeof(buf), "%s/%s", rootfs_dir, opq_marker);                                                    
    if (retval == -1 || retval >= sizeof(buf)) {                                                                             
        singularity_message(ERROR, "Error with pathname too long\n");                                                        
        ABORT(255);                                                                                                          
    }                                                                                                                        
                                                                                                                             
    if (is_dir(buf) == 0)                                                                                                    
        s_rmdir(buf);                                                                                                        
                                                                                                                             
    return 0;                                                                                                                
}

Reply:
Thanks for the pointers - embarrassing how much sensible C I've forgotten living in Python etc. Just grepped through some C code from way back when I was 17 and found lots of `char buf[MAXLINE]` etc. Need to reactivate that part of my memory somehow 😨